File: typescript/src/validator/result.ts
/**
 * The outcome of validating an access token: either a valid token (with its subject, granted scopes,
 * audience and claims) or a failure carrying a stable `error` code (see {@link ./errors}).
 */
import type { ValidationError } from "./errors.js";

export interface ValidationResult {
  valid: boolean;
  /** Stable error code when `valid` is false; `null` when valid. */
  error: ValidationError | null;
  errorDescription: string | null;
  subject: string | null;
  scopes: string[];
  audience: string[];
  claims: Record<string, unknown>;
  /** The `exp` claim (integer epoch seconds) when present. */
  expiresAt: number | null;
}

export function success(
  claims: Record<string, unknown>,
  scopes: string[],
  audience: string[],
): ValidationResult {
  const sub = claims.sub;
  const exp = claims.exp;
  return {
    valid: true,
    error: null,
    errorDescription: null,
    subject: typeof sub === "string" ? sub : null,
    scopes,
    audience,
    claims,
    expiresAt: exp == null ? null : Number(exp),
  };
}

export function failure(error: ValidationError, description?: string): ValidationResult {
  return {
    valid: false,
    error,
    errorDescription: description ?? null,
    subject: null,
    scopes: [],
    audience: [],
    claims: {},
    expiresAt: null,
  };
}

File: python/src/token_validator/result.py
"""The outcome of validating an access token."""
from __future__ import annotations


class ValidationResult:
    """Either a valid token (with its subject, granted scopes, audience and claims) or a failure carrying a
    stable ``error`` code (see :mod:`token_validator.errors`)."""

    def __init__(self, valid, *, error=None, error_description=None, subject=None,
                 scopes=None, audience=None, claims=None, expires_at=None):
        self.valid = valid
        self.error = error
        self.error_description = error_description
        self.subject = subject
        self.scopes = list(scopes or [])
        self.audience = list(audience or [])
        self.claims = dict(claims or {})
        self.expires_at = expires_at

    def __bool__(self):
        return self.valid

    @classmethod
    def success(cls, claims, scopes, audience):
        return cls(True, subject=claims.get("sub"), scopes=scopes, audience=audience,
                   claims=claims, expires_at=claims.get("exp"))

    @classmethod
    def failure(cls, error, description=None):
        return cls(False, error=error, error_description=description)

    def __repr__(self):
        if self.valid:
            return f"ValidationResult(valid, sub={self.subject!r}, scopes={self.scopes})"
        return f"ValidationResult(invalid, error={self.error!r}: {self.error_description})"
